File: vol1/ch6/LeetCode144-3.py
class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        ans=[]
        if root==None:return ans
        st=[]                   #用列表st作为栈
        p=root
        while st or p!=None:
            while p!=None:
                ans.append(p.val)
                st.append(p)
                p=p.left
            if st:
                p=st.pop()
                p=p.right
        return ans

File: vol1/ch6/LeetCode144-3.cpp
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> ans;
        if(root==NULL) return ans;
        stack<TreeNode*> st;
        TreeNode*p=root;
        while (!st.empty() || p!=NULL) {
            while(p!=NULL) {
                ans.push_back(p->val);
                st.push(p);
                p=p->left;
            }
            if(!st.empty()) {
                p=st.top(); st.pop();
                p=p->right;
            }
        }
        return ans;
    }
};

